File: javascript/secret-handshake/secret-handshake.js
//
// This is only a SKELETON file for the 'Secret Handshake' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

const SIGNALS = ["wink", "double blink", "close your eyes", "jump"];

export const secretHandshake = n => {
  if (typeof n !== "number") {
    throw new Error("Handshake must be a number");
  }

  const result = SIGNALS.filter((signal, index) => n & (1 << index));

  if (n & (1 << SIGNALS.length)) {
    result.reverse();
  }

  return result;
};

File: cpp/secret-handshake/secret_handshake.cpp
#include "secret_handshake.h"
#include <array>

namespace secret_handshake
{
const std::array<std::string, 4> SIGNALS{"wink", "double blink", "close your eyes", "jump"};

std::vector<std::string> commands(int n)
{
    std::vector<std::string> result;
    int rest{n};
    for (const auto &signal : SIGNALS)
    {
        if (rest & 1)
        {
            result.push_back(signal);
        }
        rest >>= 1;
    }

    if (n >= 16)
    {
        std::reverse(result.begin(), result.end());
    }

    return result;
}
} // namespace secret_handshake
